presto-common/src/main/java/com/facebook/presto/common/type/DistinctType.java

```diff
@@ -0,0 +1,340 @@
+
+        seenTypes.put(firstAncestor.get().getName(), firstAncestor.get());
+        while (firstAncestor.get().getSuperTypeIfAvailable().isPresent()) {
+            firstAncestor = firstAncestor.get().getSuperTypeIfAvailable();
+            seenTypes.put(firstAncestor.get().getName(), firstAncestor.get());
+        }
+
+        if (seenTypes.containsKey(secondAncestor.get().getName())) {
+            return secondAncestor;
+        }
+        seenTypes.put(secondAncestor.get().getName(), secondAncestor.get());
+
+        while (secondAncestor.get().getSuperTypeIfAvailable().isPresent()) {
+            secondAncestor = secondAncestor.get().getSuperTypeIfAvailable();
+            if (seenTypes.containsKey(secondAncestor.get().getName())) {
+                return secondAncestor;
+            }
+            seenTypes.put(secondAncestor.get().getName(), secondAncestor.get());
+        }
+
+        while (firstAncestor.isPresent() || secondAncestor.isPresent()) {
+            if (firstAncestor.isPresent() && firstAncestor.get().getSuperTypeName().isPresent()) {
+                if (seenTypes.containsKey(firstAncestor.get().getSuperTypeName().get())) {
+                    return Optional.of(seenTypes.get(firstAncestor.get().getSuperTypeName().get()));
+                }
+                firstAncestor = firstAncestor.get().getSuperTypeLoadIfNeeded();
+                seenTypes.put(firstAncestor.get().getName(), firstAncestor.get());
+            }
+
+            if (secondAncestor.isPresent() && secondAncestor.get().getSuperTypeName().isPresent()) {
+                if (seenTypes.containsKey(secondAncestor.get().getSuperTypeName().get())) {
+                    return Optional.of(seenTypes.get(secondAncestor.get().getSuperTypeName().get()));
+                }
+                secondAncestor = secondAncestor.get().getSuperTypeLoadIfNeeded();
+                seenTypes.put(secondAncestor.get().getName(), secondAncestor.get());
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+
+        DistinctType other = (DistinctType) obj;
+
+        return Objects.equals(this.userDefinedDistinctType, other.userDefinedDistinctType);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(userDefinedDistinctType);
+    }
+
+    @Override
+    public TypeSignature getTypeSignature()
+    {
+        return typeSignature;
+    }
+
+    @Override
+    public String getDisplayName()
+    {
+        return userDefinedDistinctType.getName().toString();
+    }
+
+    @Override
+    public boolean isComparable()
+    {
+        return baseType.isComparable();
+    }
+
+    @Override
+    public boolean isOrderable()
+    {
+        return userDefinedDistinctType.isOrderable();
```

Comment:
why not use baseType.isOrderable()?

Reply:
Distinct type Spec defines isOrderable trait, which is controlled by the type definition

Reply:
This is sad. But make a comment somewhere.